[1, 2단계 - 체스] 러쉬(이태훈) 미션 제출합니다.

xogns1514 opened the pull request:
안녕하세요 썬 리뷰이를 맡게 된 러쉬입니다.  이전 미션에 비해 난이도가 어렵게 느껴진 미션이었네요. 

리뷰 잘 부탁드립니다.🙇🏻‍♂️

## 미션 목표

- 초반 설계를 진행 한 뒤 tdd를 단위로 진행한다.
- 객체에 하나의 책임을 맡게 한다.

## 미션을 수행하며 궁금한점

### ChessBoard의 역할

- ChessBoard의 역할: 각 기물과 위치를 저장하고, 입력에 따라 업데이트 한다
- 문제점: 각 기물들의 움직임에 대한 많은 검증까지 책임 지고 있다.
- 원인: 기물들은 자신을 제외한 기물들의 위치 및 상태를 알 수 없다.
- 궁금한점: 외부 상황에 영향을 받는 검증 책임을 ChessBoard가 갖고 있는 것이 많은 책임을 부여받고 있다고 생각하시나요?

### PieceRole 추상화 문제

- PieceRole을 추상화한 이유: 각 Piece마다 움직임 전략이 다르다.
- 문제점: Pawn을 제외한 기물들은 같은 코드로 `canMove` 를 재정의하고 있습니다.
- 원인: Pawn을 제외한 기물은 이동 전략이 턴수, 상대 기물 위치에 따라 영향을 받지 않는다.
- 궁금한점: 필드값을 제외하면, `canMove` 가 같은 코드로 재정의되고 있습니다. 서로 다른 이동 전략을 저장하고 있는 이유로 추상화를 적용한 것을 적절한 추상화라고 볼 수 있을까요?

Mingyum-Kim committed:
docs: 기능 구현 목록 작성

Mingyum-Kim committed:
feat(Rank): 체스판의 가로줄을 생성하는 기능 구현

xogns1514 committed:
feat(ChessBoard): 체스판을 생성하는 기능 구현

xogns1514 committed:
feat(Position): 칸의 위치를 생성하는 기능 구현

Mingyum-Kim committed:
feat(Square): 체스판의 한 칸을 생성하고 동등성을 비교하는 기능 구현

xogns1514 committed:
feat(Piece): 기물을 나타내는 도메인을 생성

xogns1514 committed:
feat(PieceType): 기물의 종류를 나타내는 도메인 생성

xogns1514 committed:
feat(Piece): 기물의 종류와 색깔을 포함하는 도메인 생성

xogns1514 committed:
feat(Rank): Rank 필드값을 List<Square>로 변경

Mingyum-Kim committed:
refactor(Square): PieceType이 아닌 Piece를 필드로 가지도록 수정

Mingyum-Kim committed:
feat(RankGenerator): 체스판을 초기화하는 기능 구현

xogns1514 committed:
feat(OutputView): 체스판을 출력하는 기능 구현

Mingyum-Kim committed:
feat(ChessController): 체스를 출력하는 함수를 호출하는 컨트롤러 구현

Mingyum-Kim committed:
refactor(OutputView): 레터링과 넘버링을 제외하여 출력하도록 수정

Mingyum-Kim committed:
feat(InputView): 체스 게임을 시작하는 명령어를 받고 출력하는 기능 구현

xogns1514 committed:
style: 문자열 포맷팅

xogns1514 committed:
refactor(PieceType): PieceType 클래스명 PieceRole로 변경

xogns1514 committed:
refactor(Piece): Piece 클래스명 PieceType으로 변경

xogns1514 committed:
chore: 패키지 분리

xogns1514 committed:
feat(RankGenerator): generatePieceType 메서드 가독성 개선

Mingyum-Kim committed:
docs: 2단계 미션 기능 구현 목록 작성

Mingyum-Kim committed:
Merge branch 'step1' of https://github.com/Mingyum-Kim/java-chess into step1

Mingyum-Kim committed:
feat(Piece): 기물의 유형과 위치를 저장하는 Piece 객체 구현

xogns1514 committed:
feat(RankGenerator): generate 메서드 Piece 의존성 제거

xogns1514 committed:
feat(Square): PieceType 필드 제거

xogns1514 committed:
feat(PieceMover): piece가 source에 위치한 piece를 target으로 이동하는 기능 구현

xogns1514 committed:
feat(OutputView): 불필요한 메서드 제거

Mingyum-Kim committed:
feat(PieceGenerator): 기물을 초기화하는 기능 구현

Mingyum-Kim committed:
feat(OutputView): PieceMover의 정보를 기반으로 체스판의 상태를 출력하도록 구현

xogns1514 committed:
feat(Position): 체스판의 row와 column을 File, Rank 클래스로 래핑

Mingyum-Kim committed:
feat(PieceMover): 기물을 이동시킬 때 생길 수 있는 예외를 처리

Mingyum-Kim committed:
test(PieceMover): 에러 상황에 대한 테스트 코드 작성

xogns1514 committed:
feat(Direction): source와 target을 통해 이동할 방향을 찾는 기능 구현

xogns1514 committed:
feat(King): King의 움직임 수 있는 방향을 판단하는 기능 구현

xogns1514 committed:
fix: 에러나는 부분 주석 처리

Mingyum-Kim committed:
feat(PieceRole): Piece의 움직임 방향을 정의

xogns1514 committed:
refactor: 변경된 Piece에 따라 체스판 출력하도록 수정

Mingyum-Kim committed:
docs: 기능 구현 목록 작성

Mingyum-Kim committed:
feat(ChessController): 기물을 이동하고 체스판을 출력하는 기능 구현

xogns1514 committed:
test: 기물의 움직임 가능 여부 테스트

Mingyum-Kim committed:
feat(Pawn): 폰이 이동 가능한 경로를 조회하는 기능 구현

Mingyum-Kim committed:
refactor(Direction): Direction의 결정 조건을 올바르게 수정

xogns1514 committed:
test: 폰의 움직임 가능 여부 테스트

Mingyum-Kim committed:
feat(Pawn): Pawn가 움직일 때 첫 번째 턴에서만 두 칸 이동할 수 있도록 설정

xogns1514 committed:
feat(PieceMover): 나이트를 제외한 기물은 이동경로에 기물이 있으면 이동불가 기능 구현

xogns1514 committed:
feat(ChessController): 올바르지 않은 move 명령어 예외처리

xogns1514 committed:
refactor: 코드 가독성 개선

xogns1514 committed:
refactor(PieceMover): 기물 이동 예외 메서드 분리

xogns1514 committed:
refactor(OutputView): 체스판 출력 메서드 분리

xogns1514 committed:
refactor(PieceType): Piece의 Position을 제거하고 PieceType을 Piece로 변경

xogns1514 committed:
refactor(Movable): 기물의 최대 이동 횟수를 검증하는 메서드 분리

xogns1514 committed:
refactor(Pawn): Pawn 색깔에 따라 이동경로 분리

xogns1514 committed:
refactor(PieceMover): PieceMover 클래스명 ChessBoard로 변경

Mingyum-Kim committed:
refactor(PieceGenerator): 가로 단위의 Piece를 초기화하는 로직 개선

Mingyum-Kim committed:
refactor(Vector): 방향을 결정하는 file, rank의 차이 도메인을 Vector라고 지정

Mingyum-Kim committed:
refactor: 디렉터리 구조 변경

Mingyum-Kim committed:
refactor(ChessController): 컨트롤러 메서드의 indent depth 감소

Mingyum-Kim committed:
feat(OutputView): 체스 게임을 시작할 때 메시지를 출력

Mingyum-Kim committed:
refactor(Square): Square를 Record로 구현

syoun602 commented on src/main/java/domain/piece/PieceGenerator.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    public static void generate(final ChessBoard mover) {
+        for (int row = CHESS_BOARD_SIZE; row >= 1; row--) {
```
loop를 거꾸로 돌리는 이유가 있나요??

xogns1514 replied:
[반영 완료]
출력 부분에서 영향을 받는다고 착각했습니다. 오름차순으로 변경했습니다.

syoun602 commented on src/main/java/domain/piece/PieceGenerator.java:
```diff
@@ -0,0 +1,64 @@
+
+    private static List<Piece> generateListPiece(final List<PieceRole> pieceRoles, final Color color) {
+        return pieceRoles.stream()
+                .map(pieceRole -> new Piece(pieceRole, color)).toList();
```
```suggestion
                .map(pieceRole -> new Piece(pieceRole, color))
                .toList();
```

syoun602 commented on src/main/java/domain/piece/PieceGenerator.java:
```diff
@@ -0,0 +1,64 @@
+        rankPieces.put(7, generateListPiece(FRONT_BLACK, Color.BLACK));
+        rankPieces.put(2, generateListPiece(FRONT_WHITE, Color.WHITE));
+        rankPieces.put(1, generateListPiece(BACK, Color.WHITE));
+    }
```
이름들이 한눈에 보이질 않는 것 같아요 😢 특수 기물, 폰 과 같은 명칭을 써볼 수 있지 않을까요?

xogns1514 replied:
[반영완료]
다음 이름으로 변경했습니다. 
`SPECIAL_PIECES`, `BLACK_PAWN_PIECES`

syoun602 commented on src/main/java/domain/game/Square.java:
```diff
@@ -0,0 +1,19 @@
+import domain.position.Position;
+import java.util.Objects;
+
+public record Square(Position position) {
```
Vector랑 Square의 역할의 차이가 있을까요? 🤔

xogns1514 replied:
- 기존 Square는 Position과 동일하다고 판단하여 삭제했습니다.
- ChessVector는 두 Position을 통해 만들어지는 방향성과 관련된 객체입니다.
- Position 클래스는 체스판 위의 위치를 담당하는 객체입니다.
- 두 클래스가 가진 상태는 비슷하나, 역할은 다르다고 판단하여 두 객체를 분리하게 되었습니다.

syoun602 commented on src/main/java/controller/ChessController.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    private boolean isNotValidCoordinateInput(final String input) {
+        return !Pattern.compile(REGEX_FORMAT).matcher(input).matches();
```
Pattern은 생성 비용이 비싼 객체입니다 🙂
객체 자체를 상수로두고 이름을 부여해볼 수 있겠네요

xogns1514 replied:
[반영완료] 
Pattern을 상수화했습니다.

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -0,0 +1,148 @@
+
+    private final Map<Square, Piece> pieceBySquare;
+
+    public ChessBoard() {
```
보드 생성 당시에 내부적으로 initialize를 해줄 수 있지 않을까요?

xogns1514 replied:
[반영완료]
Board 생성자 내부에서 Generator를 통해 초기화 하도록 수정했습니다.

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -0,0 +1,148 @@
+        moveValidateExceptKnight(sourceSquare, targetSquare, findPiece);
+
+        if (findPiece.canMove(sourceSquare.position(), targetSquare.position())) {
+            update(sourceSquare, targetSquare, findPiece);
```
적절한 메서드명을 찾아봅시다

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -0,0 +1,148 @@
+    private void update(Square sourceSquare, Square targetSquare, Piece findPiece) {
+        if (pieceBySquare.containsKey(targetSquare)) {
+            pieceBySquare.remove(targetSquare);
+        }
```
remove 내부 동작을 살펴볼까요?

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -0,0 +1,148 @@
+        return false;
+    }
+
+    private void pawnMoveValidate(Square sourceSquare, Square targetSquare, Piece findPiece) {
```
pawn에 대한 제어가 보드에서 일어나고 있습니다
추상화를 이용해 폰을 구분지어서 리팩토링을 진행해볼 수 있을 것 같아요 😄

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -0,0 +1,148 @@
+        if (findPiece.isNotKnight()) {
+            Direction direction = Direction.findDirection(sourceSquare.position(), targetSquare.position());
+
+            Position here = new Position(sourceSquare.position());
```
개인적으로 here 보다는 position이 더 나을 것 같습니다

syoun602 commented on src/main/java/domain/game/Direction.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    private static boolean isSameDirection(final Direction direction, final Vector vector) {
+        for (int step = 1; step <= 8; step++) {
```
이건 무슨 숫자일까요? 👀

syoun602 commented on src/main/java/domain/game/Direction.java:
```diff
@@ -0,0 +1,69 @@
+            if (canReach(findDistance(direction.fileVector, step), vector.file()) && canReach(
+                    findDistance(direction.rankVector, step), vector.rank())) {
+                return true;
+            }
```
distance에 대한 변수를 추출해 가독성을 늘려봅시다
이런 부분들은 코드 전체적으로 확인부탁드려요!

syoun602 commented on src/main/java/controller/ChessController.java:
```diff
@@ -0,0 +1,75 @@
+        this.outputView = outputView;
+    }
+
+    public void run() {
```
턴에 대한 개념이 없는 것 같아요. 다음 리뷰때까지 구현해봅시다 :)
![image](https://github.com/woowacourse/java-chess/assets/67397679/2ee8e30e-e4b8-4b51-bff3-4b1c3b432747)

syoun602 commented on src/main/java/domain/piece/piecerole/Queen.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.List;
+import java.util.Objects;
+
+public class Queen implements PieceRole {
```
퀸의 움직임은 4방향인가요?

xogns1514 replied:
[반영완료]
대각선 이동방향 추가했습니다.

syoun602 commented on src/main/java/domain/piece/piecerole/Pawn.java:
```diff
@@ -0,0 +1,68 @@
+            return 2;
+        }
+        return 1;
+    }
```
count 변수로는 코드 맥락 파악이 힘듭니다!

xogns1514 replied:
[반영완료]
`moveCount` 로 변경했습니다.

syoun602 commented on src/test/java/fixture/PositionFixture.java:
```diff
@@ -0,0 +1,33 @@
+import domain.position.Rank;
+
+public final class PositionFixture {
+    public static Position generateC1Position() {
```
사용할때 기준으로 `Position.generate...()`이라면 Position을 굳이 명시할 필요는 없을 것 같습니다~

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -0,0 +1,148 @@
+import java.util.List;
+import java.util.Map;
+
+public class ChessBoard {
```
> ChessBoard 역할

말씀하신대로 현재 구조에서는 보드에서 검증하는게 맞는 것 같습니다 😊

syoun602 commented on src/main/java/domain/piece/piecerole/PieceRole.java:
```diff
@@ -0,0 +1,8 @@
+
+import domain.position.Position;
+
+public interface PieceRole {
```
넵 적절한 추상화로 보여집니다 👍

syoun602 reviewed (changes requested):
안녕하세요 러쉬~ 리뷰어 썬입니다 ☀️

어려운 미션임에도 잘 구현해주셨네요 👍
미완성인 부분들이 보여서 리뷰 및 코멘트 남겨두었으니 확인 부탁드려요!
궁금한 점이 있으면 언제든 DM이나 코멘트를 활용해주세요 😊

xogns1514 committed:
refactor(ChessController): Pattern 객체 상수화

xogns1514 committed:
refactor(ChessBoard): ChessBoard 생성시 초기화하도록 수정

xogns1514 committed:
refactor(ChessBoard): 검증 및 업데이트 메서드명 변경

- moveValidateExceptKnight -> sameColorPieceOnRouteValidate
- update -> updateChessBoardAfterMove

xogns1514 committed:
refactor(ChessBoard):here 변수명 position으로 변경

xogns1514 committed:
refactor(Direction): 매직 넘버 상수화

- MIN_BOARD_DISTANCE: 1
- MAX_BOARD_DISTANCE: 8

xogns1514 committed:
refactor(ChessController): Pattern 상수화

xogns1514 committed:
refactor(Direction): 가독성 개선

xogns1514 committed:
refactor(Pawn): 움직임 변수명 변경

- count -> moveCount

xogns1514 committed:
fix(Queen): Queen 대각선 방향성 추가

xogns1514 committed:
refactor(Square): Square 클래스 제거

xogns1514 committed:
refactor(Direction): 움직이는 방향 찾는 메서드 수정

xogns1514 committed:
refactor(Position): Position 불변 클래스로 수정

xogns1514 committed:
refactor(PositionFixture): PositionFixture 메서드명 수정 및 static import

xogns1514 committed:
refactor: 접근 제한자 private으로 수정

xogns1514 committed:
refactor(OutputView): 파라미터명 수정

- mover -> chessBoard

xogns1514 committed:
refactor(ChessBoard): 기물의 움직임 검증 로직 인터페이스화

xogns1514 committed:
refactor(ChessGame): 턴에 따라 체스 게임 진행 도메인 구현 및 Command 패턴 적용

xogns1514 committed:
fix(ChessBoard): 자신의 기물패를 확인하기 전에 source에 기물이 있는지 검증

xogns1514 commented:
안녕하세요 썬☀️ 지난 리뷰를 바탕으로 리팩토링 진행했습니다. 특히 기존 chessBoard의 코드가 읽기 어렵다고 판단하여, 많이 수정했습니다. 2차 피드백도 잘 부탁드립니다🙇🏻‍♂️  남은 주말 잘 보내세요!

---

## 바뀐점

### 각 기물의 움직임 검증 추상화

> pawn에 대한 제어가 보드에서 일어나고 있습니다
> 
> 
> 추상화를 이용해 폰을 구분지어서 리팩토링을 진행해볼 수 있을 것 같아요 😄
> 
- 궁금한점: 이 피드백에 대해 자세히 이해하지 못했어요. 기존 Piece의 구현체인 Pawn을 BlackPawn, WhitePawn으로 구분지어 추상화할 수 있다는 의미였을까요? 이번 리팩토링에서 잘 반영되었는지 궁금합니다.
- 반영한 부분: 기존에는 Board에서 Black, White에 따라 검증을 했습니다. 리팩토링 이후, 색에 상관없이 통합으로 검증하도록 했습니다. 또한 움직임 검증을 추상화하여 `여러칸 움직일 수 있는 기물`, `한칸씩 움직이는 기물`, `Pawn`  이 세가지의 검증으로 나누었습니다.

### Command 패턴 적용

- 기존 controller에서 command에 따라 분기처리가 복잡하다고 생각했습니다.
- Command를 추상화하여, 각 Command 구현체에 따라 체스게임을 진행하도록 변경했습니다.

### 턴제 도입

- 체스게임을 진행하는 ChessGame 클래스를 통해, 플레이어 Color에 따라 기물을 움직일 수 있도록 수정했습니다.

xogns1514 commented:
안녕하세요 썬☀️ 지난 리뷰를 바탕으로 리팩토링 진행했습니다. 특히 기존 chessBoard의 코드가 읽기 어렵다고 판단하여, 많이 수정했습니다. 2차 피드백도 잘 부탁드립니다🙇🏻‍♂️  남은 주말 잘 보내세요!

---

## 바뀐점

### 각 기물의 움직임 검증 추상화

> pawn에 대한 제어가 보드에서 일어나고 있습니다
> 
> 
> 추상화를 이용해 폰을 구분지어서 리팩토링을 진행해볼 수 있을 것 같아요 😄
> 
- 궁금한점: 이 피드백에 대해 자세히 이해하지 못했어요. 기존 Piece의 구현체인 Pawn을 BlackPawn, WhitePawn으로 구분지어 추상화할 수 있다는 의미였을까요?
- 반영한 부분: 기존에는 Board에서 Black, White에 따라 따러 검증을 했습니다. 리팩토링 이후, 색에 상관없이 통합으로 검증하도록 했습니다. 또한 움직임 검증을 추상화하여 `여러칸 움직일 수 있는 기물`, `한칸씩 움직이는 기물`, `Pawn`  이 세가지의 검증으로 나누었습니다.

### Command 패턴 적용

- 기존 controller에서 command에 따라 분기처리가 복잡하다고 생각했습니다.
- Command를 추상화하여, 각 Command 구현체에 따라 체스게임을 진행하도록 변경했습니다.

### 턴제 도입

- 체스게임을 진행하는 ChessGame 클래스를 통해, 플레이어 Color에 따라 기물을 움직일 수 있도록 수정했습니다.

xogns1514 committed:
refactor(ChessBoard): 예외처리 상태 수정

xogns1514 committed:
refactor(State): State 클래스명 변경

xogns1514 committed:
refactor(ChessGame): 이미 게임 시작된 이후 start 입력시 예외처리 수정

xogns1514 committed:
refactor(ChessGame): read-only 체스판 반환하도록 변경

syoun602 commented on src/main/java/controller/ChessController.java:
```diff
@@ -1,21 +1,12 @@
 import view.InputView;
 import view.OutputView;
 
 public class ChessController {
```
구현하신 체스로 게임을 진행해보는건 어떨까요?
현재도 제대로 구동되지 않는 것 같습니다 😢
![image](https://github.com/woowacourse/java-chess/assets/67397679/b70aa554-659a-4c09-831e-5ac603ac9c44)

xogns1514 replied:
[반영완료]
Pawn이 대각선으로 상대 기물을 못잡는 버그가 있었네요. 
수정했습니다.!

syoun602 commented on src/main/java/domain/command/CommandParser.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.function.Function;
+import view.CommandType;
+
+public class CommandParser {
```
Parser 클래스를 살펴보면 단순히 enum을 한번 더 감싼것으로 보이는데, `CommandType` 내부에 함수로 들고 있는 것은 어떨까요?

xogns1514 replied:
[반영완료]
구현당시는 view의 input에 따라 Command를 parse하는 역할을 따로할 객체가 필요하다고 생각했습니다.
말씀 들어보니, CommandType과 CommandParser가 동일한 역할을 하고 있다고 생각이 드네요. 
CommandType으로 옮겼습니다.

syoun602 commented on src/main/java/domain/command/EndCommand.java:
```diff
@@ -0,0 +1,21 @@
+    @Override
+    public void execute(Execute execute) {
+        execute.end();
+    }
```
`execute(execute)`는 조금 어색한 것 같아요. 인터페이스명에 동사를 사용하기보다는 Runnable처럼 Executable로 만들어볼 수 있을 것 같네요 😄

xogns1514 replied:
[반영완료]
Excutable로 수정했습니다.

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -1,148 +1,157 @@
+
+    public void add(Position position, Piece piece) {
+        pieceByPosition.put(position, piece);
+    }
```
해당 메서드는 테스트에서만 사용되고 있습니다. 테스트를 위한 비즈니스 코드 생성은 어떤 단점들이 있을까요?

xogns1514 replied:
- 의도하지 않은 곳에서 비즈니스 해당 로직이 사용되어, 잘못된 결과를 발생시킬 수 있습니다.
- 해당 비즈니스 코드가 변경되었을 떄, 테스트 코드를 신뢰할 수 없습니다.
- 도메인 스펙에 불필요한 코드가 작성됩니다.
해당 코드를 제거하고, checkRoute 메서드를 통해 검증을 했습니다!

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -1,148 +1,157 @@
-        forwardValidate(forward, direction, targetSquare);
-        diagonalValidate(diagonals, direction, targetSquare);
+    private void validateSingleMoveRoute(Position source, Position target) {
+        checkPieceMove(source, target);
```
이 메서드는 inline해도 충분할 것 같아요

xogns1514 replied:
[반영완료]
해당 메서드를 삭제하고, checkPieceMove를 사용하도록 수정했습니다.

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -1,148 +1,157 @@
-        return pieceBySquare.get(targetSquare);
+    public Map<Position, Piece> getPieceByPosition() {
+        return Collections.unmodifiableMap(pieceByPosition);
     }
```
`board.getPieceByPosition()`은 position을 입력받아 Piece를 받아오는 메서드로 읽힐 것 같아요

xogns1514 replied:
[반영완료]
getPiecesPosition으로 메서드명 변경했습니다.

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -1,148 +1,157 @@
+    private final Map<Position, Piece> pieceByPosition;
+    private final Map<PieceRole, RouteValidator> routeValidator;
 
     public ChessBoard() {
```
사실 routeValidator가 오히려 piece를 getter로 꺼내와서 체스보드에서 검증을 진행하고 있는 것 같은데, 이 부분은 각각의 Piece가 책임질 수 있도록 리팩토링 해볼 수 있을 것 같아요!

xogns1514 replied:
[반영완료]
- 각 piece의 방향성을 검증 
- piece에게 pieceType을 물어봄
- pieceType에 따라, 외부 상황에 영향을 받는 검증 수행

Map 형태로 어떤 기물에 대한 검증인지 저장하지 않고, 기물에게 역할을 물어보고 검증을 수행하도록 변경했습니다.

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -1,148 +1,157 @@
 
-    public boolean hasPiece(final Square square) {
-        return pieceBySquare.containsKey(square);
+    public boolean isNotEmptyAt(final Position position) {
```
일관성을 지켜주세요
```suggestion
    public boolean isNotEmptyAt(Position position) {
```

syoun602 commented on src/main/java/domain/game/ChessGame.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public void start() {
+        if (gameState == GameState.RUNNING) {
```
```suggestion
        if (gameState.isRunning()) {
```
과 같은 메서드를 통해 객체에게 물어볼까요?

xogns1514 replied:
[반영완료]
State에게 메시지를 보내는 형식으로 수정했습니다.

syoun602 commented on src/main/java/domain/command/MoveCommand.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    private void checkArgumentCount(List<String> arguments) {
+        if (arguments.size() != 2) {
```
2는 어떤 숫자일까요?~

xogns1514 replied:
move 커맨드로부터 입력받아야하는 인자 갯수를 뜻하는 의미로
`MOVE_ARGUMENTS_SIZE` 로 상수화 했습니다. 이 네이밍이 의도를 드러내지 못한다고 생각하시면 알려주세요!

syoun602 commented on src/main/java/domain/move/ChessVector.java:
```diff
@@ -0,0 +1,47 @@
+
+    private int gcd() {
+        return BigInteger.valueOf(this.fileVector).gcd(BigInteger.valueOf(rankVector)).intValue();
+    }
```
알고리즘적 접근법이네요 👍 direction과 매칭시키기 위해 사용한걸로 이해했습니다~!
코드를 작성할때 최대한 주석을 안달려고 하지만, 이런 경우라면 어떤 이유로 최대공약수를 사용했는지 정도는 남기는 것도 괜찮을 것 같습니다. 너무 갑작스럽게 ChessVector를 호출해 gcd를 구하는게 다른 개발자들에게 와닿지 않는 것 같아요!

xogns1514 replied:
작성 당시에도, 수학적인 의미가 담긴 코드라 페어와 걱정했습니다..ㅎㅎ
이런 부분에는 주석을 달아놓을 필요가 있군요. 참고하겠습니다!

syoun602 commented on src/main/java/domain/move/Direction.java:
```diff
@@ -14,7 +15,6 @@ public enum Direction {
     NORTH_WEST(-1, 1),
 
-    // KNIGHT DIRECTION
     UP_RIGHT(1, 2),
```
UpRight가 위로 한칸 우로 두칸을 충분히 표현하지 못하는것 같아요
차라리 나침반을 적절히 활용해보는건 어떨까요?
![image](https://github.com/woowacourse/java-chess/assets/67397679/32159bb6-d1bd-4790-9607-bd05d5e0b6cd)

xogns1514 replied:
[반영완료]
나침반을 기준으로 방향성 이름들을 변경했습니다.

syoun602 commented on src/main/java/domain/piece/Piece.java:
```diff
@@ -19,25 +17,21 @@ public boolean isEqualColor(final Color target) {
     }
 
+    public boolean isNotEqualColor(final Color target) {
+        return this.color != target;}
```
```suggestion
        return this.color.isNotSame(target);
    }
```
컨벤션 외로, 파라미터는 target인데 Color가 오고 있습니다! 외부에서 getter를 사용하고 있겠네요 👀

xogns1514 replied:
기존 isEqualColor의 경우 외부에서 getter를 사용하고 있어, Piece를 파라미터로 받아 비교하도록 변경했습니다.
isNotEqualColor의 경우 사용자 턴을 위해 사용된 메서드로, Color를 파라미터로 받아도록 변경했습니다. 
[궁금한점]
두 메서드명을 비교하면, 같은 타입의 파라미터를 받을 것처럼 예상됩니다. 
메서드명을 수정하는것이 나을까요?

syoun602 replied:
넵 말씀하신대로 오해의 여지가 있을 수 있겠네요 👀
메서드 오버로딩으로 볼 수도 있고 컴파일 타임에 바로 알 수 있기 때문에 
메서드로 적절히 표현할 수 있다면 변경하고 아니라면 현재도 괜찮아보입니다 😊

syoun602 commented on src/main/java/domain/piece/piecerole/Knight.java:
```diff
@@ -1,21 +1,21 @@
 public class Knight implements PieceRole {
-    public static final int MAX_MOVEMENT = 1;
+    private static final int MAX_MOVEMENT = 1;
     private final List<Movable> routes;
```
routes를 상수로 만들기 vs 필드로 두고 객체 생성시 초기화하기 중 현재 방식을 선택하신 이유가 궁금합니다~

xogns1514 replied:
routes는 객체가 가진 상태라고 생각해 static으로 두지 않았습니다. 필드 초기화는 생성자 내에서 하는 저의 규칙에 따라, 생성자내에서 초기화 했습니다. 모든 Knight는 같은 route를 갖는다는 것을 생각해보면, static으로 초기화 해주어도 될 것 같네요ㅎㅎ

syoun602 commented on src/main/java/domain/piece/piecerole/Pawn.java:
```diff
@@ -1,36 +1,36 @@
+    private List<Movable> routes;
+    private int moveCount;
 
     public Pawn(final Color color) {
```
넵 말씀드린대로 Pawn을 두고 흰색, 검정색으로 나누어볼 수 있을 것 같아서요! 
현재 폰의 생성자만보더라도 분리할 경우 훨씬 생성자의 역할을 잘 수행할 것 같아요 😁
이렇게되면 moveCount를 굳이 둘 필요없이 pawn의 현재 위치가 초기 위치가 맞는지만 판단해도 되겠네요!

xogns1514 replied:
[반영완료]
기존 Pawn을 abstract class로 두고, BlackPawn, WhitePawn으로 상속받아 구현했습니다. 
또한 sourcePosition 위치에 따라 이동 전략을 수행하도록 수정했습니다.

syoun602 commented on src/main/java/view/OutputFormat.java:
```diff
@@ -0,0 +1,64 @@
+
+    public String parseChessBoard(final Map<Position, Piece> chessBoard) {
+        StringBuilder stringBuilder = new StringBuilder();
+        for (int rank = 8; rank >= 1; rank--) {
```
러쉬가 세워둔 매직넘버에 대한 기준이 따로 있으실까요?

xogns1514 replied:
도메인 특성을 의미하는 숫자, 검증해야 하는 부분 등 의미있는 이름을 부여할 수 있는 수를 상수화합니다.
꼼꼼히 확인하지 못한 부분이 많네요..ㅎㅎ 수정하겠습니다!

syoun602 commented on src/main/java/view/OutputFormat.java:
```diff
@@ -0,0 +1,64 @@
+    private String parsePositionByFile(Map<Position, Piece> chessBoard, int rank) {
+        StringBuilder stringBuilder = new StringBuilder();
+        for (int file = 0; file < 8; file++) {
+            Position position = new Position(new Position(new File((char) ('a' + file)), new Rank(rank)));
```
가독성이 떨어집니다 😢

xogns1514 replied:
[반영완료]
file과 rank 부분 분리하여 가독성을 개선했습니다.

syoun602 reviewed (changes requested):
안녕하세요 러쉬! 
빠르게 리뷰 반영해주셨네요 👍
조금 더 다듬어볼 부분들이 있는 것 같아서 RC로 두었습니다~!
혹시 코멘트에 대한 답변을 빠르게 받아보고 싶다면 DM 주셔도 됩니다 😊

xogns1514 committed:
test(ChessBoardTest): IllegalArgumentException 검사하도록 수정

xogns1514 committed:
fix(ChessBoard): pawn 대각선으로 상대 기물 못잡는 버그 수정

xogns1514 committed:
refactor(CommandParser): CommandParser 로직 CommandType으로 이동 및 클래스 삭제

xogns1514 committed:
refactor(Executable): Execute 클래스명 수정 및 가독성 개선을 위한 CommandConstructor 생성

xogns1514 committed:
style: 문자열 포맷팅

xogns1514 committed:
refactor(MoveCommand): 매직넘버 상수화

xogns1514 committed:
refactor(ChessBoard): validateSingeMoveRoute 삭제

xogns1514 committed:
refactor(GameState): 게임이 진행중인 상태인지 확인하는 기능 추가

xogns1514 committed:
chore(ChessVector): Vector 메서드에 대한 주석 추가

xogns1514 committed:
refactor(Direction): Direction 방향명 나침반 기준으로 변경

xogns1514 committed:
refactor(Piece): 색깔 일치 유무 메서드명 변경 및 파라미터 수정

xogns1514 committed:
refactor: Pawn 제외 기물들 각 route 상수화

xogns1514 committed:
refactor(Pawn): Pawn추상화하고 WhitePawn과 BlackPawn을 각각 구현

xogns1514 committed:
refactor(ChessBoard): 각 PieceRole에 따라 검증 수행하도록 수정

xogns1514 committed:
refactor(OutputFormat): 가독성 개선

xogns1514 committed:
refactor(ChessBoard): getter 메서드명 변경

xogns1514 committed:
chore(move): move 패키지명 변경

xogns1514 committed:
test: File, Rank 경계 테스트 추가 및 Command 테스트 추가

xogns1514 committed:
test(CommandType): start, end command 예외처리 테스트 추가

xogns1514 commented:
## 피드백 반영

안녕하세요 썬☀️ 지난번 피드백 참고해서 리팩토링 진행했습니다. 또 부족한 점 있다면 말씀해주세요. 궁금한 부분은 code에 기록해두겠습니다. 잘 부탁드립니다.🙇🏻‍♂️

---

## 바뀐점

### Pawn 구분

- Pawn을 추상화하여 WhitePawn, BlackPawn으로 구분했습니다.
- 이유: 각각의 역할을 구분하고, 이동가능한 움직임을 분리하기 위해서 입니다.

### ChessGame 각 기물에 대한 움직임 검증

- 기존의 기물 별로 검증해야할 메서드를 저장한 Map을 제거했습니다. 외부 기물 위치에 따른 검증은, 해당되는 기물인지 묻고, 수행하도록 변경했습니다.
- 이유: ChessBoard에는 각 기물에 대해 어떤  검증을 수행할 책임이 없기 때문입니다.

xogns1514 commented on src/main/java/domain/piece/piecerole/PieceRole.java:
```diff
@@ -0,0 +1,12 @@
+
+    boolean canMove(Position sourcePosition, Position targetPosition);
+
+    boolean isPawn();
+
+    boolean isSlidingPiece();
```
Pawn과 SlidingPiece(비숍, 퀸, 룩)에게는 외부 기물에 따른 검증을 수행해야합니다.
검증을 수행하기 위해 pieceType을 Piece에게 물어봅니다. 
궁금한점: PieceRole은 추상화된 클래스인데, 어떤 기물인지 검사하기 위해 이러한 메서드를 두어도 적절하다고 할 수 있나요? isInstance를 직접적으로 사용하는 것을 피하기 위해 작성해두었는데, 다른점이 있는지 궁금합니다

syoun602 replied:
말씀하신대로 instanceof를 지양하기 위해 불가피하게 추가해야할 메서드일 것 같아요. 
현재 방식은 다형성을 적절히 활용한 방식임으로 문제가 없어보입니다! 
이전에 블랙잭에서 보셨을 지도 모르지만 [instanceof의 사용을 지양하자](https://tecoble.techcourse.co.kr/post/2021-04-26-instanceof/)를 한번 읽어보시면 좋을 것 같네요 😊

syoun602 commented on gradlew.bat:
```diff
@@ -41,7 +41,7 @@ if defined JAVA_HOME goto findJavaFromJavaHome
 
```
요 파일은 잘못 변경된 것 같아요~

xogns1514 replied:
[반영완료]
원래 파일로 변경했습니다.

syoun602 commented on src/main/java/domain/command/MoveCommand.java:
```diff
@@ -1,21 +1,24 @@
 import java.util.regex.Pattern;
 
 public class MoveCommand implements Command {
     private static final String MOVE_COMMAND_REGEX_FORMAT = "^[a-h][1-8]";
```
상수도 public > private 순으로 선언해주세요 !

xogns1514 replied:
[반영완료]
public -> private 순서로 변경했습니다.

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -1,57 +1,39 @@
-        pieceByPosition.put(position, piece);
+    public void checkRoute(Position source, Position target, Color color) {
+        checkSourcePosition(source);
+        checkTargetPosition(source, target);
```
메서드 내부적으로 `isSamePosition(source, target)` 메서드는 단순한 `source.equals(target)` 로 두어도 될 것 같습니다

xogns1514 replied:
[반영완료]
isSamePosition 메서드를 제거하고, 바로 source.equals(target)으로 비교하도록 수정했습니다.

syoun602 commented on src/main/java/domain/game/ChessGame.java:
```diff
@@ -21,32 +21,31 @@ public boolean isNotEnd() {
+        chessBoard.checkRoute(source, target, color);
         chessBoard.move(source, target);
 
         color = Color.reverseColor(color);
```
static 메서일 필요가 없어보입니다 👀

xogns1514 replied:
[반영완료]
color 인스턴스 한테 반대 색을 반환하도록 물어볼 수 있겠네요.
일반 메서드로 변경했습니다.

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -81,28 +63,17 @@ private boolean isSamePosition(Position sourcePosition, Position targetPosition)
-    private void validateMultipleMoveRoute(Position source, Position target) {
-        checkPieceMove(source, target);
-        validateOtherPieceOnRoute(source, target);
+    private void validateSlidingMove(Position source, Position target, Piece piece) {
+        if (piece.isSlidingPiece()) {
+            validateOtherPieceOnRoute(source, target);
+        }
     }
```
지금도 충분히 괜찮지만, 이부분과 `validatePawnMove()`를 제거하는 방향으로 조금 더 고민해봐도 좋을 것 같습니다 😊

syoun602 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -134,24 +109,32 @@ private void validatePawnMove(Position source, Position target) {
+    public void move(Position source, Position target) {
+        Piece findPiece = piecesPosition.get(source);
+
+        piecesPosition.remove(target);
+        piecesPosition.put(target, findPiece);
+        piecesPosition.remove(source);
```
요렇게도 가능하겠네요~
```suggestion
        piecesPosition.put(target, findPiece);
        piecesPosition.remove(source);
```

xogns1514 replied:
[반영완료]
put할때 업데이트 되니 remove는 필요없네요ㅎㅎ

syoun602 commented on src/main/java/domain/movement/Direction.java:
```diff
@@ -33,11 +33,11 @@ public enum Direction {
 
     public boolean isNorthOrSouth() {
-        return this == Direction.NORTH || this == Direction.SOUTH;
+        return this == Direction.N || this == Direction.S;
```
```suggestion
        return this == N || this == S;
```
같은 클래스 내라면 클래스명을 붙일 필요가 없습니다 ~

xogns1514 replied:
[반영완료]
오 그렇군요..!
Direction을 붙이지 않는 것이 가독성이 더 좋아 보이네요

syoun602 commented on src/main/java/domain/piece/piecerole/BlackPawn.java:
```diff
@@ -0,0 +1,55 @@
+public class BlackPawn extends Pawn {
+    private static final int INITIAL_MAX_MOVEMENT = 2;
+    private static final int GENERAL_MAX_MOVEMENT = 1;
+    private static final List<Movable> INITIAL_ROUTES = List.of(
+            new Movable(INITIAL_MAX_MOVEMENT, Direction.S),
+            new Movable(GENERAL_MAX_MOVEMENT, Direction.SE),
+            new Movable(GENERAL_MAX_MOVEMENT, Direction.SW)
+    );
+    private static final List<Movable> GENERAL_ROUTES = List.of(
+            new Movable(GENERAL_MAX_MOVEMENT, Direction.S),
+            new Movable(GENERAL_MAX_MOVEMENT, Direction.SE),
+            new Movable(GENERAL_MAX_MOVEMENT, Direction.SW)
+    );
```
중복도 있고, 한 눈에 GENERAL_MAX_MOVEMENT인지 INITIAL인지 구분이 되지 않습니다!
하나만 두는 방법으로 변경해보는건 어떤가요?

xogns1514 replied:
[반영완료]
Route를 하나로 합쳤습니다.
초기 위치인 경우 Movable에게 최대 거리를 물어보고, filtering 하는 방법으로 수정했습니다.

syoun602 commented on src/main/java/domain/piece/piecerole/Pawn.java:
```diff
@@ -1,68 +1,10 @@
-    @Override
-    public int hashCode() {
-        return Objects.hash(routes);
+        return true;
```
사용될 일이 없어 보이는 메서드입니다!

xogns1514 replied:
[반영완료]
앗 하위 클래스에서 구현하고 있다면 비어두게 둘 수 있군요. 삭제하겠습니다.

syoun602 commented on src/main/java/domain/piece/piecerole/Rook.java:
```diff
@@ -1,33 +1,38 @@
+    @Override
+    public boolean isPawn() {
+        return false;
     }
 
     @Override
-    public boolean canMove(final Position sourcePosition, final Position targetPosition) {
-        return routes.stream()
-                .anyMatch(movable -> movable.canMove(sourcePosition, targetPosition));
+    public boolean isSlidingPiece() {
+        return true;
```
추상화의 단계를 적절히 더 쪼개보면 이런 중복도 줄여볼 수 있겠네요 😁

xogns1514 replied:
[반영완료]
상위에 SlidingPiece와 NonSlidingPiece를 두어 추상화를 한단계 추가할 수 있겠네요!!
아이디어 감사합니다ㅎㅎ

syoun602 commented on src/main/java/domain/position/File.java:
```diff
@@ -3,12 +3,21 @@
 
+    private void validateLetterRange(char letter) {
+        if (letter < MIN_LETTER || letter > MAX_LETTER) {
+            throw new IllegalArgumentException();
```
예외 메시지는 없나요?~

xogns1514 replied:
[반영완료]
예외 메시지를 빠뜨렸네요..! 추가했습니다.

syoun602 commented on src/main/java/domain/position/Position.java:
```diff
@@ -31,6 +27,14 @@ public ChessVector toVector(Position target) {
         return new ChessVector(file.subtract(target.file), rank.subtract(target.rank));
     }
 
+    public boolean isAtRank2() {
+        return rank.isRank2();
+    }
+
+    public boolean isAtRank7() {
+        return rank.isRank7();
+    }
+
```
2랑 7은 어떤 의미가 있는지 한번 생각해보게 될 것 같아요 🤔
`isAtRank(int number)`라는 메서드를 만들고 폰에서 자신의 초기 위치를 넘겨주어 판단하도록 두는건 어떨까요?

xogns1514 replied:
[반영완료]
rank에게 인자로 넘겨준 뒤 확인하는것이, 역할 분리가 되겠네요.
수정했습니다.!

syoun602 reviewed (approved):
안녕하세요 러쉬!

체스는 복잡한 도메인인만큼 구현하는 방식이 여러가지가 있는데 잘 구현해주셨네요 💯
이번 단계에서 시도해볼 부분들은 대부분 진행한 것 같아 이만 머지하겠습니다 !
간단한 코멘트 남겨두었으니 다음 단계 진행하며 함께 반영해주세요~ 

다음 단계도 화이팅입니다 💪
